crypto/evp/digest.c

```diff
@@ -628,34 +641,30 @@ static void *evp_md_from_dispatch(const OSSL_DISPATCH *fns,
             if (md->dupctx == NULL)
                 md->dupctx = OSSL_get_OP_digest_dupctx(fns);
             break;
-        case OSSL_FUNC_DIGEST_SIZE:
-            if (md->size == NULL)
-                md->size = OSSL_get_OP_digest_size(fns);
-            break;
-        case OSSL_FUNC_DIGEST_BLOCK_SIZE:
-            if (md->dblock_size == NULL)
-                md->dblock_size = OSSL_get_OP_digest_block_size(fns);
-            break;
-        case OSSL_FUNC_DIGEST_SET_PARAMS:
-            if (md->set_params == NULL)
-                md->set_params = OSSL_get_OP_digest_set_params(fns);
-            break;
         case OSSL_FUNC_DIGEST_GET_PARAMS:
             if (md->get_params == NULL)
                 md->get_params = OSSL_get_OP_digest_get_params(fns);
             break;
+        case OSSL_FUNC_DIGEST_CTX_SET_PARAMS:
+            if (md->ctx_set_params == NULL)
+                md->ctx_set_params = OSSL_get_OP_digest_ctx_set_params(fns);
+            break;
+        case OSSL_FUNC_DIGEST_CTX_GET_PARAMS:
+            if (md->ctx_get_params == NULL)
+                md->ctx_get_params = OSSL_get_OP_digest_ctx_get_params(fns);
+            break;
         }
     }
     if ((fncnt != 0 && fncnt != 5)
-        || (fncnt == 0 && md->digest == NULL)
-        || md->size == NULL) {
+        || (fncnt == 0 && md->digest == NULL)) {
         /*
          * In order to be a consistent set of functions we either need the
          * whole set of init/update/final etc functions or none of them.
          * The "digest" function can standalone. We at least need one way to
          * generate digests.
          */
         EVP_MD_meth_free(md);
+        EVPerr(0, EVP_R_INVALID_PROVIDER_FUNCTIONS);
```

Comment:
Agreed.  `EVPerror` anyone?

Reply:
I think if you add new macros you should do complete replacement and deprecate the existing ones.  Otherwise a mix of new and old macros will be inconsistent and even more ugly.  I am willing to work on a set of PR's that does this if there is interest.

Reply:
I agree that the 0 is ugly.

Reply:
> I am willing to work on a set of PR's that does this if there is interest.

I'm interested in seeing that (with some trepidation since I know its going to cause havoc with back porting).

Reply:
One PR would be sufficient :grinning:

```
$ grep -lr err.*_F_ . | wc -l
     399
$ grep -r err.*_F_ . | wc
    3619
```

The back porting is significant concern.  I suspect it's worth doing at some point.  Maybe timed with the next LTS release?

Reply:
See #9441